tree_arena/src/tree_arena_unsafe.rs

```diff
@@ -0,0 +1,617 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(unsafe_code, reason = "Purpose is unsafe abstraction")]
+use super::NodeId;
+
+use std::cell::UnsafeCell;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+struct TreeNode<T> {
+    item: T,
+    children: Vec<NodeId>,
+}
+
+/// Mapping of data for the Tree Arena
+#[derive(Debug)]
+struct DataMap<T> {
+    /// The items in the tree
+    items: HashMap<NodeId, Box<UnsafeCell<TreeNode<T>>>>,
+    /// The parent of each node, or None if it is the root
+    parents: HashMap<NodeId, Option<NodeId>>,
+    /// The roots of the tree
+    roots: Box<UnsafeCell<Vec<NodeId>>>,
```

Comment:
On further thought, I observed the problem regarding inconsistency with mem::swap when the tests were in the same file (and so accessed the internals) - I think the encapsulation is enough to prevent this as the roots array and the DataMap always belong to the same underlying tree